util/cli.c

```diff
@@ -0,0 +1,430 @@
+ * @param [in] dev - Pointer to the device descriptor.
+ * @param [in] str - Pointer to the string to be displayed.
+ * @return 0 in case of success, error code otherwise.
+ */
+int32_t cli_write_string(struct cli_desc *dev, uint8_t *str)
+{
+	uart_write_nonblocking(dev->uart_device, str, strlen((char *)str));
+
+	while(!uart_written);
+	uart_written = 0;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief UART callback specific to UART.
+ * @param app_param - Application supplied CLI handler.
+ * @param event - Interrupt code.
+ * @param data - Not used here.
+ */
+void cli_uart_callback(void *app_param, uint32_t event,
+		       void *data)
+{
+	struct cli_desc *cli_ptr = app_param;
+
+	switch(event) {
+	case READ_DONE:
+		cli_parse(cli_ptr, uart_buff);
+		uart_read_nonblocking(cli_ptr->uart_device, &uart_buff, 1);
+		break;
+	case WRITE_DONE:
+		uart_written = 1;
+		break;
+	case ERROR:
+		break;
+	}
+}
+
+/**
+ * @brief Setup the CLI module of the application. Allocate memory for the
+ *        driver handler and initialize the physical layer.
+ * @param [out] device    - Pointer to the device handler structure.
+ * @param [in] init_param - Pointer to the device initialization structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t cli_setup(struct cli_desc **device, struct cli_init_param *init_param)
+{
+	int32_t ret;
+
+	*device = (struct cli_desc *)calloc(1, sizeof(**device));
+	if (!(*device))
+		return -1;
+
+	ret = uart_init(&((*device)->uart_device), &(init_param->uart_init));
+	if (ret != SUCCESS)
+		goto error_dev;
+
+	ret = uart_read_nonblocking((*device)->uart_device, &uart_buff, 1);
+	if (ret != SUCCESS)
+		goto error_uart;
+
+	return SUCCESS;
+
+error_uart:
+	uart_remove((*device)->uart_device);
+error_dev:
+	free(*device);
+
+	return FAILURE;
+}
+
+/**
+ * @brief Free resources allocated by cli_setup().
+ * @param [in] dev - Pointer to the device descriptor.
+ * @return 0 if success, negative error code otherwise.
+ */
+int32_t cli_remove(struct cli_desc *dev)
+{
+	int32_t ret;
+
+	if(!dev)
+		return -1;
+
+	ret = uart_remove(dev->uart_device);
+	if(ret != SUCCESS)
+		return ret;
+
+	free(dev);
+
+	return ret;
+}
+
+/**
+ * @brief Implements CLI feedback.
+ * @param [in] dev - The device structure.
+ * @param [in] c   - The newest character to be parsed.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t cli_parse(struct cli_desc *dev, uint8_t c)
+{
+	int8_t i;
+
+	switch(c) {
+	case _CR:
+	case _LF:
+		uart_cmd = 1;
+		break;
+	case _BS:
+		if(uart_line_index == 0)
+			break;
+		uart_line_index--;
+		break;
+	case _TB:
+		cli_write_string(dev, (uint8_t*)"\n>");
+		i = 0;
+		do {
+			uart_current_line[i] = uart_previous_line[i];
+		} while(uart_previous_line[i++] != '\0');
+		cli_write_string(dev, uart_current_line);
+		uart_line_index = --i;
+		break;
+	case _NC:
+		break;
+	default:
+		uart_current_line[uart_line_index++] = c;
+		if(uart_line_index == 256) {
```

Comment:
are you sure this is the correct condition ? uart_line_index is uint8_t

Reply:
I think this is a mistake. Will fix.